### ISSUE TITLE
proposal: remove cached type of instructions

### ISSUE THREAD

**Issue by mewmew** [llir/llvm#116]
The proposal of this issue is to remove the result type cache of instructions in the [ir](https://godoc.org/github.com/llir/llvm/ir) package. Or at least investigate the benefits and drawbacks of removing the type cache. First, lets start with a background of what the type cache is, before getting into discussing if we should remove or keep it.

For most value instructions (i.e. [ir.Instruction](https://godoc.org/github.com/llir/llvm/ir#Instruction)s implementing the [value.Value](https://godoc.org/github.com/llir/llvm/ir/value#Value) interface), the `type` of the resulting value computed by the instruction is currently cached.

To give an example, lets consider the [InstXor](https://godoc.org/github.com/llir/llvm/ir#InstXor) instruction.

```go
type InstXor struct {
    // Name of local variable associated with the result.
    LocalIdent
    // Operands.
    X, Y value.Value // integer scalars or vectors

    // Type of result produced by the instruction.
    Typ types.Type
    // (optional) Metadata.
    Metadata
}
```

The result type is cached in the [Typ](https://godoc.org/github.com/llir/llvm/ir#InstXor.Typ) field of the `InstXor` struct, and the cached type is computed as follows:

From: https://github.com/llir/llvm/blob/b85ca175725ff59102503c86be2934bd3b9a99d3/ir/inst_bitwise.go#L328
```go
func (inst *InstXor) Type() types.Type {
	// Cache type if not present.
	if inst.Typ == nil {
		inst.Typ = inst.X.Type()
	}
	return inst.Typ
}
```

If we were to remove the result type cache of `InstXor`, the `Type()` method would look as follows:

```go
func (inst *InstXor) Type() types.Type {
	return inst.X.Type()
}
```

The result type cache was introduced primarily as a (potentially premature) optimization to avoid memory allocation when invoking the `Type` method for instructions that need to compute their result type.

The issue with caching the result type is that if you were to modify a given IR and change say the type of a function parameter from `i32` to `i64`, then this change would propagate correctly for all uses of the function parameter, but if the parameter was used as the parameter of an instruction (say an `xor` instruction), then the result type of the `xor` instruction would have a cached `i32` type, ever after updating the type of the function parameter to `i64`. To recompute the type, a user of the `ir` package would then have to reset the `Typ` field of `InstXor` to `nil`, so that the next invocation of `Type()` would recompute the cache.

This approach is indeed doable, although fragile and perhaps unintuitive, at least to users not firmly aware of the internal details of result type caching.

For now, I've added a [ResetTypes](https://godoc.org/github.com/llir/irutil#ResetTypes) convenience function to the `irutil` package which resets the type cache of value instructions. This function is analogous to the [ResetNames](https://godoc.org/github.com/llir/irutil#ResetNames) function which is used to reset the IDs of unnamed local identifiers in functions.

While introducing the `ResetTypes` function makes handling reset of type caches less painful, there are still subtle bugs that may exist for users who does not know of its existence and thus forget to call `ResetTypes` where appropriate.

I propose that we investigate removing cached result types of value instructions altogether. Potentially keeping one edge case where we cache the result of the `getelementptr` instruction, as essentially all this function does is to compute a result type. With this rationale, we may also wish to keep the cached type of `call` instructions and the `invoke` and `callbr` terminators, since they record information about the callee/invokee which may otherwise be unavailable.

The potential downside of removing the type cache is that we may get more memory allocations if the `Type()` method is invoked a lot. However, if that ever were to become an issue, then tight loops invoking `Type` could introduce a purpose specific type cache should the need arise.

The benefit of removing the type cache is a less confusing and more intuitive `ir` API.

Food for thought.

Cheers,
Robin

cc: @pwaller

**Comment by dannypsnl** [comment]
It sounds more like the owner(in the case is a function instruction) didn't
update the instructions inside of It. But correctly updating is
complicated. I think we can use a big enough Codebase to measure how much
performance would get impact without cache. If a typical project which uses
llir didn't get a lot of effects from this, then I think we can say caching
is not a typical case, and without it makes the project easier to maintain,
it's worth to remove type caching.

**Comment by mewmew** [comment]
> I think we can use a big enough Codebase to measure how much
performance would get impact without cache. f a typical project which uses
llir didn't get a lot of effects from this, then I think we can say caching
is not a typical case, and without it makes the project easier to maintain,
it's worth to remove type caching.

Sounds good. I'm currently evaluating removing the type cache, and perform transformations on some 2'056'524 lines of LLVM IR :)

**Comment by mewmew** [comment]
Gotta love self-referential instructions like these.

```llvm
; PR12075
define i1 @infinite_gep() {
  ret i1 1

unreachableblock:
  %X = getelementptr i32, i32 *%X, i32 1
  %Y = icmp eq i32* %X, null
  ret i1 %Y
}
```

Oh, and before I forget. While working on assessing if we could remove the cached type, a bug was uncovered in the current computation of `gep` result type:

```diff
diff --git a/ir/inst_memory.go b/ir/inst_memory.go
index 3cd4131..f9acac1 100644
--- a/ir/inst_memory.go
+++ b/ir/inst_memory.go
@@ -549,9 +498,10 @@ func gepInstType(elemType, src types.Type, indices []value.Value) types.Type {
        var idxs []gep.Index
        for _, index := range indices {
                var idx gep.Index
-               if index, ok := index.(constant.Constant); ok {
+               switch index := index.(type) {
+               case constant.Constant:
                        idx = getIndex(index)
-               } else {
+               default:
                        idx = gep.Index{HasVal: false}
                        // Check if index is of vector type.
                        if indexType, ok := index.Type().(*types.VectorType); ok {
```

Prior to this diff, `index` would always be false in the `else`-branch since it was type asserted to be `constant.Constant` and failed.

**Comment by mewmew** [comment]
So, for reference, besides the pre-mature optimization aspect of cached result types, the type cache was useful as they helped solve infinite recursions of Type computation like these (which apparently is valid LLVM IR):

```llvm
define void @f() {
	%x = add i32 %x, %x
	ret void
}
```

**Comment by mewmew** [comment]
Given that self-referential values are allowed in LLVM IR, I don't know of a good/possible way to solve type resolution without caching at least *some* of the result types computed by instructions.

Example of self-referential LLVM IR instruction.
```llvm
	%x = add i32 %x, %x
```

In other words, if we define the `Type()` method of `InstAdd` as follows:

```go
func (inst *InstAdd) Type() types.Type {
   return inst.X.Type()
}
```

We would not have to cache the result type of the instruction, as it would be inferred by the type of its *x* operand. (A semantic pass would ensure that *x* and *y* have the same type.)

However, if the type of *x* is the type of the instruction (as is the case with self-referential instructions), then this leads to an infinite recursion.

I don't see any obvious way to solve this issue.

Open for ideas!

cc: @dannypsnl, @pwaller.

**Comment by dannypsnl** [comment]
Infinite recursion is not acceptable since it directly makes code unusable.
So the way it should go seems like was let type would be updated when we
update the sub-instruction. A possible idea is comparing the cached type
with the computed type.

Here we can write:

```go

func (inst *InstAdd) Type() types.Type {

   if inst.Typ != inst.X.Type() { // I thought we can compare `nil`
without problem?

      inst.Typ = inst.X.Type()

   }
   return inst.Typ
}

```

And for all instructions then updating is not an issue I thought.

**Comment by mewmew** [comment]
> A possible idea is comparing the cached type
with the computed type.

This could be a good approach! Thanks for the idea.

**Comment by mewmew** [comment]
Following the principle of *don't be clever* I'd prefer to either not use a type cache at all, or to use a "dumb" type cache that does not try to update itself. At least in this way, the results are predicable and the user knows what to do when they perform custom updates of values (and their types); i.e. reset the type-cache with `inst.Typ = nil`.

@dannypsnl, I like the approach you presented to handle recalculation of the type-cache when dealing with cyclic references. However, even so, the recalculation could lead to stack overflows (as to check that the type-cache type has not changed, we'd need to check the type of the instruction, and when dealing with cyclic references in instructions, then this type would have to invoke `Type`, ad infinitum, recursively). Also, the benefit of having a type cache is that we don't have to create new types whenever we invoke the `Type` method. But if we intend to check whether the type-cache is correct when invoking `Type`, we have to create the *current*  type each time we invoke `Type` to compare against the type-cache`.

So, in summary. Lets stick to the simple approach of having a "dumb" type-cache.

Cheers,
Robin

**Comment by dannypsnl** [comment]
@mewmew Do you mind a stupid question? I want to know how can we generate code like:

```llvm
%x = add i32 %x, %x
```

by current API, I have no idea how to make it, unless I do some wired code:

```go
	instAdd := &InstAdd{LocalIdent: LocalIdent{}}
	instAdd.X = instAdd
	instAdd.Y = instAdd
	fmt.Printf("%s\n", instAdd.Type())
```

**Comment by mewmew** [comment]
> by current API, I have no idea how to make it, unless I do some wired code:
>
> ```go
>	instAdd := &InstAdd{LocalIdent: LocalIdent{}}
>	instAdd.X = instAdd
>	instAdd.Y = instAdd
>	fmt.Printf("%s\n", instAdd.Type())
> ```

Indeed, that's how you'd have to do it, as it's a cyclic-reference instruction.

Another way to do it is simply to use `asm.ParseFile`/`asm.ParseString`:

```go
package main

import (
	"fmt"
	"log"

	"github.com/llir/llvm/asm"
)

const source = `
define i32 @foo() {
	%x = add i32 %x, %x
	ret i32 %x
}
`

func main() {
	m, err := asm.ParseString("cyclic.ll", source)
	if err != nil {
		log.Fatalf("%+v", err)
	}
	fmt.Println(m)
}
```

```
$ go run main.go
# Output:

define i32 @foo() {
; <label>:0
	%x = add i32 %x, %x
	ret i32 %x
}
```

**Comment by dannypsnl** [comment]
I suddenly realized if we distinguish left, right value, then removing cache would be fine. How? First, in current implementation any instructions were a value can be using by others instructions, than compute recursively was impossible since self reference might existed. What if non-recursive from construction? Instruction is not a value, but it’s label is, then instruction to instruction where self reference occurred were eliminated.

**Comment by dannypsnl** [comment]
What would that effect current API, for example, assign instruction to instruction was invalid:
instAdd.X = inst (wrong)
But
instAdd.X = inst.ToValue()
Hence, we need to introduce new interface: Valueable for ToValue method

**Comment by mewmew** [comment]
> Instruction is not a value, but it’s label is.

Hi @dannypsnl. I very much appreciate that you keep thinking about different solutions and continue to invest effort into this issue. I understand the desire to simplify the implementation of `llir/llvm/ir` by removing the type cache. Especially as it can cause subtle type errors if modifying the IR programatically without corresponding update/reset of the type cache.

However, having an instruction be a value is a *very strong property* to have in the IR. It enables fundamental and very powerful APIs, such as the [IR walker](https://github.com/llir/irutil/blob/af3d8eb90900a7fe409aca3c97bd31325ee15e24/walk.go#L27) which allows both arbitrary inspection of the IR, but also arbitrary modification; even of [operand values as expressed by their underlying instruction](https://github.com/llir/irutil/blob/af3d8eb90900a7fe409aca3c97bd31325ee15e24/walk.go#L407).

This strength of the API is what made Go fix possible. And indeed, the IR walker implementation is based on the Go fix source code, at least in spirit.

So, for what it is worth. I personally believe that we can mark this issue as an unfortunate side-effect, but also be happy in that our current IR architecture enable powerful IR walker APIs and similar `value.Value` manipulations.

Cheers,
Robin

**Comment by dannypsnl** [comment]
I thought I can make a draft implementation for this, with the mutual reference I believed we could save both benefits of current/described-above implementation, once I had time...QQ.
The key point was the current implementation also suffer by updating order dependency, hence, this was not the problem for mixed implementation(I mean still have the same problem).

**Comment by mewmew** [comment]
> I thought I can make a draft implementation for this, with the mutual reference I believed we could save both benefits of current/described-above implementation, once I had time...QQ.

Sounds like a good idea. Feel free to prepare a draft implementation once you have time and decide to. Then we can evaluate the benefits and drawbacks of both approaches and take a decision on what to use.

With you good winter holidays @dannypsnl!

Cheers,
Robin
